File: java/src/model/HorizontalMove.java
package model;

import logger.Logger;
import exceptions.IllegalMoveException;

/**
 * @author Jakub Rak
 */
class HorizontalMove extends Move {
	private final Model model;
	private final Piece piece;
	private Piece capturedPiece;
	private final Square startSquare;
	private final Square endSquare;
	private final boolean wasMoved;
	private HorizontalMove castling;

	HorizontalMove(final Model model, final Piece piece, final Square startSquare, final Square endSquare) {
		this.model = model;
		this.piece = piece;
		capturedPiece = model.getPiece(endSquare);
		this.startSquare = startSquare;
		this.endSquare = endSquare;
		wasMoved = piece.wasMoved();
		castling = null;
	}

	/**
	 * Checks if if it might be castling move.
	 * @return true if it might be castling move, otherwise false
	 */
	boolean checkCastling() {
		if (startSquare.getFile() - endSquare.getFile() == -2 && piece.getColour() == Colour.WHITE) {
			final Piece rook = model.getPiece(new Square(7, 0));
			final HorizontalMove horizontalMove = new HorizontalMove(model, rook, new Square(7, 0), new Square(4, 0));
			if (rook != null && rook.wasMoved() == false && rook.isMoveValid(horizontalMove)) {
				castling = horizontalMove;
				return true;
			}
			return false;
		}
		else if (startSquare.getFile() - endSquare.getFile() == 2 && piece.getColour() == Colour.WHITE) {
			final Piece rook = model.getPiece(new Square(0, 0));
			final HorizontalMove horizontalMove = new HorizontalMove(model, rook, new Square(0, 0), new Square(2, 0));
			if (rook != null && rook.wasMoved() == false && rook.isMoveValid(horizontalMove)) {
				castling = horizontalMove;
				return true;
			}
			return false;
		}
		else if (startSquare.getFile() - endSquare.getFile() == -2 && piece.getColour() == Colour.BLACK) {
			final Piece rook = model.getPiece(new Square(7, 7));
			final HorizontalMove horizontalMove = new HorizontalMove(model, rook, new Square(7, 7), new Square(4, 7));
			if (rook != null && rook.wasMoved() == false && rook.isMoveValid(horizontalMove)) {
				castling = horizontalMove;
				return true;
			}
			return false;
		}
		else if (startSquare.getFile() - endSquare.getFile() == 2 && piece.getColour() == Colour.BLACK) {
			final Piece rook = model.getPiece(new Square(0, 7));
			final HorizontalMove horizontalMove = new HorizontalMove(model, rook, new Square(0, 7), new Square(2, 7));
			if (rook != null && rook.wasMoved() == false && rook.isMoveValid(horizontalMove)) {
				castling = horizontalMove;
				return true;
			}
			return false;
		}
		else
			return false;
	}

	@Override
	void doMove() throws IllegalMoveException {
		if (castling == null) {
			if (capturedPiece != null)
				model.removePieceFromList(capturedPiece);
			model.setPieceOnChessboard(endSquare, piece);
			model.setPieceOnChessboard(startSquare, null);
			piece.setCurrentSquare(endSquare);
			piece.setWasMoved(true);
		}
		else {
			model.setPieceOnChessboard(castling.endSquare, piece);
			model.setPieceOnChessboard(startSquare, null);
			piece.setCurrentSquare(castling.endSquare);
			piece.setWasMoved(true);
			if (model.isKingInCheck() == true) {
				castling = null;
				model.undoMove();
				throw new IllegalMoveException();
			}
			model.setPieceOnChessboard(endSquare, piece);
			model.setPieceOnChessboard(startSquare, null);
			piece.setCurrentSquare(endSquare);
			model.setPieceOnChessboard(castling.endSquare, castling.piece);
			model.setPieceOnChessboard(castling.startSquare, null);
			castling.piece.setCurrentSquare(castling.endSquare);
			castling.piece.setWasMoved(true);
		}
		model.changeTurn();
		if (model.isKingInCheck() == true) {
			undoMove();
			throw new IllegalMoveException();
		}
		Logger.print(this, "The move has been done.");
	}

	@Override
	Piece getCapturedPiece() {
		return capturedPiece;
	}

	@Override
	Square getEndSquare() {
		return endSquare;
	}

	@Override
	Piece getPiece() {
		return piece;
	}

	@Override
	Square getStartSquare() {
		return startSquare;
	}

	/**
	 * Checks if it is one step move.
	 * @return true if it is one step move, otherwise false
	 */
	boolean isOneStepMove() {
		return Math.abs(startSquare.getFile() - endSquare.getFile()) == 1;
	}

	@Override
	boolean isPathClear() {
		Square square = endSquare;
		if (model.isSquareFree(square) == true)
			square = nextSquare(square);
		else if (piece.getColour() != model.getPiece(square).getColour()) {
			capturedPiece = model.getPiece(square);
			square = nextSquare(square);
			Logger.print(this, "Capture..");
		}
		else {
			Logger.print(this, "Path isn't clear.");
			return false;
		}
		while (square.equals(startSquare) == false) {
			if (model.isSquareFree(square) == false) {
				Logger.print(this, "Path isn't clear.");
				return false;
			}
			square = nextSquare(square);
		}
		Logger.print(this, "The path is clear...");
		return true;
	}

	/**
	 * Gets next square on move's path (direction from {@link #endSquare} to {@link #startSquare})
	 * @param currentSquare - current square
	 * @return next square
	 */
	private Square nextSquare(final Square currentSquare) {
		final Square nextSquare = new Square();
		nextSquare.setRank(currentSquare.getRank());
		if (endSquare.getFile() < startSquare.getFile())
			nextSquare.setFile(currentSquare.getFile() + 1);
		else
			nextSquare.setFile(currentSquare.getFile() - 1);
		return nextSquare;
	}

	@Override
	void undoMove() {
		model.setPieceOnChessboard(startSquare, piece);
		model.setPieceOnChessboard(endSquare, capturedPiece);
		if (capturedPiece != null)
			model.addPieceToList(capturedPiece);
		piece.setCurrentSquare(startSquare);
		piece.setWasMoved(wasMoved);
		if (castling != null) {
			model.setPieceOnChessboard(castling.startSquare, castling.piece);
			model.setPieceOnChessboard(castling.endSquare, null);
			castling.piece.setCurrentSquare(castling.startSquare);
			castling.piece.setWasMoved(wasMoved);
		}
		model.changeTurn();
		Logger.print(this, "The move has been undone.");
	}

	@Override
	boolean wasMoved() {
		return wasMoved;
	}
}

File: c++/horizontalmove.cpp
#include "horizontalmove.h"
#include "piece.h"
#include "rook.h"
#include "game.h"
#include "illegalmoveexception.h"
#include <math.h>

namespace Model
{
    HorizontalMove::HorizontalMove(Game *game, Piece *piece, const Square &startSquare, const Square &endSquare)
    {
        this->game = game;
        this->piece = piece;
        //capturedPiece = game->getPiece(endSquare);
        this->startSquare = startSquare;
        this->endSquare = endSquare;
        capturedPiece = NULL;
        wasMoved = piece->getWasMoved();
        castling = NULL;
    }

    bool HorizontalMove::checkCastling()
    {
        if (startSquare.getFile() - endSquare.getFile() == -2 && piece->getColor() == BLACK)
        {
            Piece *rook = game->getPiece(Square(7, 0));
            HorizontalMove *horizontalMove = new HorizontalMove(game, rook, Square(7, 0), Square(4, 0));
            if (rook != NULL && !rook->getWasMoved() && rook->isMoveValid(horizontalMove))
            {
                castling = horizontalMove;
                return true;
            }
            return false;
        }
        else if (startSquare.getFile() - endSquare.getFile() == 2 && piece->getColor() == BLACK)
        {
            Piece *rook = game->getPiece(Square(0, 0));
            HorizontalMove *horizontalMove = new HorizontalMove(game, rook, Square(0, 0), Square(2, 0));
            if (rook != NULL && !rook->getWasMoved() && rook->isMoveValid(horizontalMove))
            {
                castling = horizontalMove;
                return true;
            }
            return false;
        }
        else if (startSquare.getFile() - endSquare.getFile() == -2 && piece->getColor() == WHITE)
        {
            Piece *rook = game->getPiece(Square(7, 7));
            HorizontalMove *horizontalMove = new HorizontalMove(game, rook, Square(7, 7), Square(4, 7));
            if (rook != NULL && !rook->getWasMoved() && rook->isMoveValid(horizontalMove))
            {
                castling = horizontalMove;
                return true;
            }
            return false;
        }
        else if (startSquare.getFile() - endSquare.getFile() == 2 && piece->getColor() == WHITE)
        {
            Piece *rook = game->getPiece(Square(0, 7));
            HorizontalMove *horizontalMove = new HorizontalMove(game, rook, Square(0, 7), Square(2, 7));
            if (rook != NULL && !rook->getWasMoved() && rook->isMoveValid(horizontalMove))
            {
                castling = horizontalMove;
                return true;
            }
            return false;
        }
        else
            return false;
    }

    void HorizontalMove::doMove() /*throws IllegalMoveException*/
    {
        if (castling == NULL)
        {
            if (capturedPiece != NULL)
                game->removePieceFromList(capturedPiece);
            game->setPieceOnChessboard(endSquare, piece);
            game->setPieceOnChessboard(startSquare, NULL);
            piece->setCurrentSquare(endSquare);
            piece->setWasMoved(true);
        }
        else
        {
            game->setPieceOnChessboard(castling->endSquare, piece);
            game->setPieceOnChessboard(startSquare, NULL);
            piece->setCurrentSquare(castling->endSquare);
            piece->setWasMoved(true);
            if (game->isKingInCheck())
            {
                castling = NULL;
                game->undoMove();
                throw IllegalMoveException();
            }
            game->setPieceOnChessboard(endSquare, piece);
            game->setPieceOnChessboard(startSquare, NULL);
            piece->setCurrentSquare(endSquare);
            game->setPieceOnChessboard(castling->endSquare, castling->piece);
            game->setPieceOnChessboard(castling->startSquare, NULL);
            castling->piece->setCurrentSquare(castling->endSquare);
            castling->piece->setWasMoved(true);
        }
        game->changeTurn();
        if (game->isKingInCheck())
        {
            undoMove();
            throw IllegalMoveException();
        }
        //Logger.print(this, "The move has been done.");
    }

    bool HorizontalMove::isOneStepMove()
    {
        return abs(startSquare.getFile() - endSquare.getFile()) == 1;
    }

    bool HorizontalMove::isPathClear()
    {
        Square square = endSquare;
        if (game->isSquareFree(square) == true)
            square = nextSquare(square);
        else if (piece->getColor() != game->getPiece(square)->getColor())
        {
            capturedPiece = game->getPiece(square);
            square = nextSquare(square);
            //Logger.print(this, "Capture..");
        }
        else {
            //Logger.print(this, "Path isn't clear.");
            return false;
        }
        while (square != startSquare)
        {
            if (!game->isSquareFree(square))
            {
                //Logger.print(this, "Path isn't clear.");
                return false;
            }
            square = nextSquare(square);
        }
        //Logger.print(this, "The path is clear...");*/
        return true;
    }

    Square HorizontalMove::nextSquare(const Square &currentSquare)
    {
        Square nextSquare;
        nextSquare.setRank(currentSquare.getRank());
        if (endSquare.getFile() < startSquare.getFile())
            nextSquare.setFile(currentSquare.getFile() + 1);
        else
            nextSquare.setFile(currentSquare.getFile() - 1);
        return nextSquare;
    }

    void HorizontalMove::undoMove()
    {
        game->setPieceOnChessboard(startSquare, piece);
        game->setPieceOnChessboard(endSquare, capturedPiece);
        if (capturedPiece != NULL)
            game->addPieceToList(capturedPiece);
        piece->setCurrentSquare(startSquare);
        piece->setWasMoved(wasMoved);
        if (castling != NULL)
        {
            game->setPieceOnChessboard(castling->startSquare, castling->piece);
            game->setPieceOnChessboard(castling->endSquare, NULL);
            castling->piece->setCurrentSquare(castling->startSquare);
            castling->piece->setWasMoved(wasMoved);
        }
        game->changeTurn();
        //Logger.print(this, "The move has been undone.");
    }
}

